lib/kfifo-test.c

```diff
@@ -0,0 +1,234 @@
+	KUNIT_EXPECT_EQ(test, out_data, 65);
+}
+
+static void kfifo_test_out_should_pop_multiple_elements(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+
+	kfifo_put(&my_fifo, 11);
+	kfifo_put(&my_fifo, 25);
+	kfifo_put(&my_fifo, 65);
+
+	u8 *out_buffer = kunit_kmalloc_array(test, 3, sizeof(u8), GFP_KERNEL);
+	int copied_elements = kfifo_out(&my_fifo, out_buffer, 3);
+	KUNIT_EXPECT_EQ(test, copied_elements, 3);
+
+	KUNIT_EXPECT_EQ(test, out_buffer[0], 11);
+	KUNIT_EXPECT_EQ(test, out_buffer[1], 25);
+	KUNIT_EXPECT_EQ(test, out_buffer[2], 65);
+	KUNIT_EXPECT_TRUE(test, kfifo_is_empty(&my_fifo));
+}
+
+static void
+kfifo_test_declare_init_should_define_an_empty_fifo(struct kunit *test)
+{
+	DECLARE_KFIFO(my_fifo, u8, 32);
+	INIT_KFIFO(my_fifo);
+
+	// my_fifo is a struct with an inplace buffer
+	KUNIT_EXPECT_FALSE(test, __is_kfifo_ptr(&my_fifo));
+
+	KUNIT_EXPECT_TRUE(test, kfifo_initialized(&my_fifo));
+}
+
+static void kfifo_test_define_should_equal_declare_init(struct kunit *test)
+{
+	// declare a variable my_fifo of type struct kfifo of u8
+	DECLARE_KFIFO(my_fifo1, u8, 32);
+	// initialize the my_fifo variable
+	INIT_KFIFO(my_fifo1);
+
+	// DEFINE_KFIFO declares the variable with the initial value
+	// essentially the same as calling DECLARE_KFIFO and INIT_KFIFO
+	DEFINE_KFIFO(my_fifo2, u8, 32);
+
+	// my_fifo1 and my_fifo2 are equal
+	KUNIT_EXPECT_EQ(test, sizeof(my_fifo1), sizeof(my_fifo2));
+	KUNIT_EXPECT_EQ(test, kfifo_initialized(&my_fifo1),
+			kfifo_initialized(&my_fifo2));
+	KUNIT_EXPECT_EQ(test, kfifo_is_empty(&my_fifo1),
+			kfifo_is_empty(&my_fifo2));
+}
+
+static void kfifo_test_alloc_should_initiliaze_a_ptr_fifo(struct kunit *test)
+{
+	DECLARE_KFIFO_PTR(my_fifo, u8);
+	INIT_KFIFO(my_fifo);
+
+	// kfifo_initialized returns false signaling the buffer pointer is NULL
+	KUNIT_EXPECT_FALSE(test, kfifo_initialized(&my_fifo));
+
+	// kfifo_alloc allocates the buffer
+	int ret = kfifo_alloc(&my_fifo, 32, GFP_KERNEL);
+	KUNIT_EXPECT_EQ_MSG(test, ret, 0, "Memory allocation should succeed");
+	KUNIT_EXPECT_TRUE(test, kfifo_initialized(&my_fifo));
+
+	// kfifo_free frees the buffer
+	kfifo_free(&my_fifo);
+}
+
+static void
+kfifo_test_peek_should_not_remove_elements_from_fifo(struct kunit *test)
+{
+	DEFINE_KFIFO(my_fifo, u8, 32);
+	u8 out_data;
+	int processed_elements;
+
+	// If the fifo is empty, peek returns 0
+	processed_elements = kfifo_peek(&my_fifo, &out_data);
+	KUNIT_EXPECT_EQ(test, processed_elements, 0);
+
+	kfifo_put(&my_fifo, 3);
+	kfifo_put(&my_fifo, 5);
+	kfifo_put(&my_fifo, 11);
+
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 3);
+
+	processed_elements = kfifo_peek(&my_fifo, &out_data);
+	KUNIT_EXPECT_GT(test, processed_elements, 0);
+	KUNIT_EXPECT_EQ(test, out_data, 3);
+
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 3);
+
+	processed_elements = kfifo_peek(&my_fifo, &out_data);
+	KUNIT_EXPECT_GT(test, processed_elements, 0);
+	KUNIT_EXPECT_EQ(test, out_data, 3);
+
+	KUNIT_EXPECT_EQ(test, kfifo_len(&my_fifo), 3);
```

Comment:
isn't this exactly the same as the previous one?

Reply:
Yes, the ideia is to show that doing the peek twice returns the same results, that is, returns the same element and doesn't alter the kfifo length

Reply:
Added